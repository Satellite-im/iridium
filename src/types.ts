import { CreateJWEOptions, CreateJWSOptions, DecryptJWEOptions } from 'dids';
import * as json from 'multiformats/codecs/json';
import { CID } from 'multiformats/cid';
import type { GeneralJWS } from 'dids';
import type { PeerId } from 'ipfs-core/ipns';
import type { IPFSConfig } from 'ipfs-core/dist/src/components/network';
import { JWE } from 'did-jwt';
import type { IPFS } from 'ipfs-core';
import { RecursivePartial } from '@libp2p/interfaces';
import { Libp2pInit } from 'libp2p';

export type IridiumLogger = {
  log: (...args: any[]) => void;
  debug: (...args: any[]) => void;
  error: (...args: any[]) => void;
  warn: (...args: any[]) => void;
  info: (...args: any[]) => void;
};

export type IridiumRequestOptions = {
  timeout?: number;
  signal?: AbortSignal;
};

export type IridiumResolveOptions = IridiumRequestOptions & {
  nocache?: boolean;
  recursive?: boolean;
};

export type IridiumReadOptions = {
  decrypt?: boolean;
  decryptOptions?: DecryptJWEOptions;
  verifySignature?: boolean;
};

export type IridiumDagGetOptions = IridiumResolveOptions & {
  path?: string;
  localResolve?: boolean;
};

export type IridiumLoadOptions = IridiumReadOptions & {
  depth?: number;
  dag?: IridiumDagGetOptions;
};

export type IridiumGetOptions = {
  load?: IridiumLoadOptions;
  resolve?: IridiumResolveOptions;
};

export type IridiumWriteOptions = IridiumRequestOptions & {
  encrypt?: IridiumEncryptOptions;
  sign?: boolean | CreateJWSOptions;
  dag?: IridiumDagPutOptions;
};

export type IridiumSendOptions = IridiumWriteOptions & {
  to: string | string[];
  encode?: boolean;
};

export type IridiumSetOptions = {
  store?: IridiumWriteOptions;
  publish?: IridiumPublishOptions;
};

export type IridiumDagPutOptions = IridiumRequestOptions & {
  storeCodec?: string;
  inputCodec?: string;
  hashAlg?: string;
  cid?: CID | string;
  pin?: boolean;
};

export type IridiumEncryptOptions = {
  recipients?: string[];
  options?: CreateJWEOptions;
};

export type IridiumPayloadBody = GeneralJWS | JWE | IridiumDocument | string;
export type IridiumPayloadJWE = {
  type: 'jwe';
  body: JWE;
};
export type IridiumPayloadJWS = {
  type: 'jws';
  body: GeneralJWS;
};
export type IridiumPayloadJSON = {
  type: 'json';
  body: IridiumDocument;
};
export type IridiumPayloadText = {
  type: 'text';
  body: string;
};
export type IridiumPayload =
  | IridiumPayloadJWE
  | IridiumPayloadJWS
  | IridiumPayloadJSON
  | IridiumPayloadText;

export type IridiumPayloadBytes = json.ByteView<IridiumPayload>;

export type IridiumPayloadEvent<B = IridiumPayload> = {
  payload: B;
};

export type IridiumMessage = {
  from: PeerId;
  to?: string | PeerId | string[];
};

export type IridiumChannelEvent<B = IridiumPayload> = IridiumPayloadEvent<B> & {
  channel: string;
  data?: IridiumDocument | string;
};

export type IridiumErrorEvent = {
  error: Error;
  message?: string;
};

export type IridiumPubsubEvent<B = IridiumPayload> = IridiumMessage & {
  topic: string;
} & { data: json.ByteView<B> };

export type IridiumPeerMessage<T = IridiumDocument | string> =
  IridiumMessage & {
    topic: string;
    payload: T;
  };

export type IridiumSyncNodeConfig = {
  label?: string;
  peerId: string;
  multiaddr?: string;
};

export type IridiumConfig = {
  repo?: string;
  version?: string;
  followedPeers?: string[];
  syncNodes?: IridiumSyncNodeConfig[];
  ipfs?: IPFSConfig;
  libp2p?: RecursivePartial<Libp2pInit>;
};

export type IridiumSeedConfig = {
  config?: IridiumConfig;
  ipfs?: IPFS;
  peerId?: PeerId;
  logger?: IridiumLogger;
};

export type IridiumDocument = {
  [key: string]: any;
};

export type IridiumPeer = {
  id: string;
  did: string;
  channel: string;
  meta: any;
  seen: number;
};

export type IridiumPublishOptions = IridiumRequestOptions & {
  resolve?: boolean;
  lifetime?: string;
  ttl?: string;
  key?: string;
  allowOffline?: boolean;
};
