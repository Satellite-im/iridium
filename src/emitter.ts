export type EmitterCallback<T> = (event: T) => void

export class Emitter<T = any> {
  private readonly _store: { [key: string]: any[] }

  constructor() {
    this._store = Object.create(null)
  }

  on<H = T>(event: string, cb: EmitterCallback<H>) {
    return this.event(event).push(cb)
  }

  off<H = T>(event: string, cb: EmitterCallback<H> | false = false) {
    const stack = this.event(event)
    if (cb === false) {
      return stack.splice(0, stack.length)
    }

    return stack.splice(stack.indexOf(cb) >>> 0, 1)
  }

  emit<H = T>(event: string, data: H, isAsync = false) {
    const result = [
      ...this.event(event).map((cb) => cb(data)),
      ...this.event('*').map((cb) =>
        cb(Object.assign(data, { channel: event })),
      ),
    ]
    return isAsync ? Promise.all(result) : result
  }

  event(named: string) {
    if (!this._store[named]) {
      this._store[named] = []
    }
    return this._store[named]
  }

  size() {
    return Array.from(Object.values(this._store), (s) => s.length).reduce(
      (a, b) => a + b,
      0,
    )
  }
}
export default Emitter
