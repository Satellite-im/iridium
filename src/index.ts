export { default as Iridium, default } from './iridium';
export { default as Emitter, type EmitterCallback } from './emitter';
export { ipfsNodeFromKey, ipfsConfig } from './ipfs';
export * from './types';
