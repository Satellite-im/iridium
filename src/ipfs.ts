import merge from 'merge';
import { createFromPrivKey } from '@libp2p/peer-id-factory';
import { create } from 'ipfs-core';
import type { IPFS } from 'ipfs-core';
import { keys } from '@libp2p/crypto';
import { IridiumConfig } from './iridium';

export async function ipfsNodeFromSeed(
  seed: Uint8Array,
  config: IridiumConfig = {}
): Promise<IPFS> {
  const key = await keys.supportedKeys.ed25519.generateKeyPairFromSeed(seed);
  const peerId = await createFromPrivKey(key);
  return create(ipfsConfig(peerId, config));
}

export function ipfsConfig(peerId: any, config: IridiumConfig = {}) {
  console.info(
    'repo',
    `${config.repo || 'iridium'}/${
      config.version || 'v1.0.0'
    }/${peerId.toString()}`
  );
  return merge(
    {
      repo: `${config.repo || 'iridium'}/${
        config.version || 'v1.0.0'
      }/${peerId.toString()}`,
      config: {
        Addresses: {
          Swarm: config.swarm,
        },
        Bootstrap: config.bootstrap || [],
        Discovery: {
          MDNS: {
            Enabled: true,
            Interval: 1,
          },
          webRTCStar: {
            Enabled: true,
          },
        },
        Pubsub: {
          PubSubRouter: 'floodsub',
          Enabled: true,
        },
        Routing: {
          type: 'dhtclient',
        },
        Swarm: {
          DisableRelay: false,
          EnableRelayHop: true,
        },
        Identity: {
          peerId,
        },
      },
      relay: {
        enabled: true,
        hop: {
          enabled: true,
          active: true,
        },
      },
      init: {
        privateKey: peerId,
        algorithm: 'ed25519',
      },
    },
    config
  );
}
