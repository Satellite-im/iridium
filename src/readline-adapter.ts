import * as json from 'multiformats/codecs/json';
import minimist from 'minimist';
import Iridium from './iridium';
import type { IridiumMessage } from './iridium';
import Emitter from './emitter';

export type IridiumTerminalCommands = {
  [key: string]: (...args: any[]) => any;
};

const textEncoder = new TextEncoder();

export default class IridiumTerminal extends Emitter<IridiumMessage> {
  private _commands: IridiumTerminalCommands;
  private _help: IridiumTerminalCommands;

  constructor(
    instance: Iridium,
    customCommands: IridiumTerminalCommands,
    help: IridiumTerminalCommands = {}
  ) {
    super();
    instance.on('*', (event: IridiumMessage) => {
      console.info(
        `[iridium] ${event.channel} ${event.from ? `(${event.from}):` : ''} ${
          event.payload ? JSON.stringify(event.payload, null, 2) : ''
        }`
      );
      this.emit(event.channel as string, event);
    });
    this._commands = {
      send: instance.send.bind(instance),
      sendSigned: instance.sendSigned.bind(instance),
      sendEncrypted: instance.sendEncrypted.bind(instance),
      broadcast: instance.broadcast.bind(instance),
      broadcastSigned: instance.broadcastSigned.bind(instance),
      load: instance.load.bind(instance),
      loadSigned: instance.loadSigned.bind(instance),
      loadEncrypted: instance.loadEncrypted.bind(instance),
      get: instance.get.bind(instance),
      set: instance.set.bind(instance),
      help: this.help.bind(this),
      whoami: () => {
        return { peerId: instance.peerId, did: instance.id };
      },
      clear: () => {
        console.clear();
      },
      ...customCommands,
    };
    this._help = help;
  }

  async exec(line: string) {
    const {
      _: [cmd, ...args],
      ...props
    } = minimist(line.split(' '));
    const command = this.command(cmd);
    const input = args.map((arg) =>
      arg.startsWith('enc:')
        ? textEncoder.encode(arg.substring(4))
        : arg.startsWith('json:')
        ? JSON.parse(arg.substring(5))
        : arg
    );
    try {
      const result = await command(...input);
      console.info(JSON.stringify(result, null, 2));
    } catch (e) {
      console.error(e);
    }
  }

  command(cmd: string) {
    return this._commands[cmd] || this._commands.help.bind(this, cmd);
  }

  help(cmd: string, ...args: any[]) {
    if (cmd) {
      return `${cmd} - ${
        this._help[cmd](...args) || 'No help available for this command'
      }`;
    } else {
      return 'Available commands: ' + Object.keys(this._commands).join(', ');
    }
  }
}
