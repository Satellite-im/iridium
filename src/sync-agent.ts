import * as json from 'multiformats/codecs/json';
import Iridium from './iridium';
import { IridiumConfig, IridiumSeedConfig } from './types';
import IridiumTerminal from './readline-adapter';
import * as readline from 'node:readline/promises';
import { stdin as input, stdout as output } from 'node:process';

export type IridiumSyncAgentConfig = {
  peers?: {
    limit?: number;
    timeout?: number;
    retry?: number;
  };
  sync?: {
    limit?: number;
    maxSize?: number;
    timeout?: number;
    retry?: number;
  };
  pins?: {
    limit?: number;
  };
  identity?: {
    limit?: number;
    maxSize?: number;
  };
  server?: IridiumSeedConfig;
};

const localRelay =
  process.env.IRIDIUM_LOCAL_RELAY ||
  process.env.NUXT_ENV_IRIDIUM_LOCAL_RELAY ||
  process.env.VITE_ENV_IRIDIUM_LOCAL_RELAY;
if (localRelay) {
  console.info(`Using local relay peer: ${localRelay}`);
}

const DEFAULT_CONFIG: IridiumSyncAgentConfig = {
  peers: {
    limit: 10000,
    timeout: 10000,
    retry: 3,
  },
  sync: {
    limit: 10000,
    timeout: 10000,
    maxSize: 256,
    retry: 3,
  },
  identity: {
    limit: 100000,
    maxSize: 256,
  },
};

type PeerData = {
  did: string;
  peerId: string;
  seen: number;
  pins: string[];
};

export default class IridiumSyncAgent {
  private _peers: { [key: string]: PeerData } = {};
  private _offline: { [key: string]: string[] } = {};
  private _pins: { [key: string]: string[] } = {};
  constructor(
    private readonly instance: Iridium,
    private config: IridiumSyncAgentConfig
  ) {
    this.instance.on('peer:connect', this.onPeerConnect.bind(this));
    this.instance.on(`sync/${this.instance.id}`, this.onSyncMessage.bind(this));
  }

  async start() {
    const terminal = new IridiumTerminal(this.instance, {});
    await this.instance.start();
    await terminal.exec('whoami');
    await terminal.exec('pins');
    await this.instance.subscribe(`sync/${this.instance.id}`);
    await this.instance.subscribe(`friends/announce`);

    const rl = readline.createInterface({ input, output });
    rl.setPrompt('> ');

    rl.on('line', async (line: string) => {
      await terminal.exec(line).catch((error) => {
        console.error(error);
      });
      rl.prompt();
    }).on('close', () => {
      console.info('bye');
      process.exit(0);
    });

    console.info('sync agent terminal starting');
    rl.prompt();
  }

  static async create(
    seed: string,
    config: IridiumSyncAgentConfig = DEFAULT_CONFIG
  ) {
    const iridium = await Iridium.fromSeedString(seed, config.server);
    const agent = new IridiumSyncAgent(iridium, config);
    await agent.start();
    return agent;
  }

  onPeerConnect({ peerId }: { peerId: string }) {
    console.info('peer connected', peerId);
  }

  async onSyncMessage(message: any) {
    console.info(message);
    const { from, payload, type } = message;

    if (!['jwe', 'jws'].includes(type)) {
      console.debug('ignoring unsigned sync message', message);
      return;
    }

    const remotePeerId = from.toString();
    const remoteDID = await Iridium.peerIdToDID(from);
    if (payload.type === 'sync-init') {
      this._peers[remotePeerId] = {
        did: remoteDID,
        peerId: remotePeerId,
        seen: Date.now(),
        pins: payload.data?.pins || [],
      };
      await this.instance.send(
        {
          action: 'sync-init',
          data: {
            request: payload.request || undefined,
            success: true,
          },
        },
        {
          to: remotePeerId,
          encrypt: { recipients: [remoteDID] },
        }
      );
      return;
    }

    if (payload.type === 'sync-put') {
      if (!this._offline[remotePeerId]) {
        this._offline[remotePeerId] = [];
      }
      if (
        this._offline[remotePeerId].length >= Number(this.config.sync?.limit)
      ) {
        throw new Error('offline sync limit reached for peer: ' + remotePeerId);
      }
      // pin the document until it can be delivered
      const cid = await this.instance.store(payload, {
        encrypt: { recipients: [this.instance.id, remoteDID, payload.to] },
        dag: { pin: true },
      });
      this._offline[remotePeerId].push(cid.toString());
      await this.instance.send(
        {
          action: 'sync-put',
          data: {
            cid,
            request: payload.request || undefined,
            success: true,
          },
        },
        {
          to: remotePeerId,
          encrypt: { recipients: [remoteDID] },
        }
      );
    }

    if (payload.type === 'pin') {
      if (!this._pins[remotePeerId]) {
        this._pins[remotePeerId] = [];
      }

      if (this._pins[remotePeerId].length >= Number(this.config.pins?.limit)) {
        throw new Error('pin limit reached for peer: ' + remotePeerId);
      }

      try {
        const cid = payload.data;
        await this.instance.ipfs.pin.add(cid);
        await this.instance.send(
          {
            action: 'pin',
            data: {
              cid,
              request: payload.request || undefined,
              success: true,
            },
          },
          {
            to: remotePeerId,
            encrypt: { recipients: [remoteDID] },
          }
        );
      } catch (error) {
        console.error(error);
      }
    }

    console.warn('unhandled sync message', message);
  }
}
