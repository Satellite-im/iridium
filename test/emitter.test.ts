import * as emitter from "../src/emitter"

describe("emitter.Emitter.emit", () => {
    let inst3: any
    let inst: any
    let inst2: any

    beforeEach(() => {
        inst3 = new emitter.Emitter()
        inst = new emitter.Emitter()
        inst2 = new emitter.Emitter()
    })

    test("0", () => {
        let result: any = inst2.emit("data:image/svg+xml;charset=UTF-8,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20version%3D%221.1%22%20baseProfile%3D%22full%22%20width%3D%22undefined%22%20height%3D%22undefined%22%3E%3Crect%20width%3D%22100%25%22%20height%3D%22100%25%22%20fill%3D%22grey%22%2F%3E%3Ctext%20x%3D%22NaN%22%20y%3D%22NaN%22%20font-size%3D%2220%22%20alignment-baseline%3D%22middle%22%20text-anchor%3D%22middle%22%20fill%3D%22white%22%3Eundefinedxundefined%3C%2Ftext%3E%3C%2Fsvg%3E", undefined, true)
        expect(result).toMatchSnapshot()
    })

    test("1", () => {
        let result: any = inst.emit("data:image/svg+xml;charset=UTF-8,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20version%3D%221.1%22%20baseProfile%3D%22full%22%20width%3D%22undefined%22%20height%3D%22undefined%22%3E%3Crect%20width%3D%22100%25%22%20height%3D%22100%25%22%20fill%3D%22grey%22%2F%3E%3Ctext%20x%3D%22NaN%22%20y%3D%22NaN%22%20font-size%3D%2220%22%20alignment-baseline%3D%22middle%22%20text-anchor%3D%22middle%22%20fill%3D%22white%22%3Eundefinedxundefined%3C%2Ftext%3E%3C%2Fsvg%3E", undefined, false)
        expect(result).toMatchSnapshot()
    })

    test("2", () => {
        let result: any = inst3.emit("", undefined, false)
        expect(result).toMatchSnapshot()
    })
})

describe("emitter.Emitter.off", () => {
    let inst4: any
    let inst3: any
    let inst: any
    let inst2: any

    beforeEach(() => {
        inst4 = new emitter.Emitter()
        inst3 = new emitter.Emitter()
        inst = new emitter.Emitter()
        inst2 = new emitter.Emitter()
    })

    test("0", () => {
        let result: any = inst2.off("data:image/svg+xml;charset=UTF-8,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20version%3D%221.1%22%20baseProfile%3D%22full%22%20width%3D%22undefined%22%20height%3D%22undefined%22%3E%3Crect%20width%3D%22100%25%22%20height%3D%22100%25%22%20fill%3D%22grey%22%2F%3E%3Ctext%20x%3D%22NaN%22%20y%3D%22NaN%22%20font-size%3D%2220%22%20alignment-baseline%3D%22middle%22%20text-anchor%3D%22middle%22%20fill%3D%22white%22%3Eundefinedxundefined%3C%2Ftext%3E%3C%2Fsvg%3E", false)
        expect(result).toMatchSnapshot()
    })

    test("1", () => {
        let result: any = inst.off("data:image/svg+xml;charset=UTF-8,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20version%3D%221.1%22%20baseProfile%3D%22full%22%20width%3D%22undefined%22%20height%3D%22undefined%22%3E%3Crect%20width%3D%22100%25%22%20height%3D%22100%25%22%20fill%3D%22grey%22%2F%3E%3Ctext%20x%3D%22NaN%22%20y%3D%22NaN%22%20font-size%3D%2220%22%20alignment-baseline%3D%22middle%22%20text-anchor%3D%22middle%22%20fill%3D%22white%22%3Eundefinedxundefined%3C%2Ftext%3E%3C%2Fsvg%3E", true)
        expect(result).toMatchSnapshot()
    })

    test("2", () => {
        let result: any = inst3.off("", false)
        expect(result).toMatchSnapshot()
    })
})

describe("emitter.Emitter.on", () => {
    let inst: any
    let inst2: any

    beforeEach(() => {
        inst = new emitter.Emitter()
        inst2 = new emitter.Emitter()
    })

    test("0", () => {
        let result: any = inst2.on("data:image/svg+xml;charset=UTF-8,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20version%3D%221.1%22%20baseProfile%3D%22full%22%20width%3D%22undefined%22%20height%3D%22undefined%22%3E%3Crect%20width%3D%22100%25%22%20height%3D%22100%25%22%20fill%3D%22grey%22%2F%3E%3Ctext%20x%3D%22NaN%22%20y%3D%22NaN%22%20font-size%3D%2220%22%20alignment-baseline%3D%22middle%22%20text-anchor%3D%22middle%22%20fill%3D%22white%22%3Eundefinedxundefined%3C%2Ftext%3E%3C%2Fsvg%3E", () => undefined)
        expect(result).toMatchSnapshot()
    })

    test("1", () => {
        let result: any = inst.on("", () => undefined)
        expect(result).toMatchSnapshot()
    })
})

describe("emitter.Emitter.event", () => {
    let inst5: any
    let inst4: any
    let inst3: any
    let inst: any
    let inst2: any

    beforeEach(() => {
        inst5 = new emitter.Emitter()
        inst4 = new emitter.Emitter()
        inst3 = new emitter.Emitter()
        inst = new emitter.Emitter()
        inst2 = new emitter.Emitter()
    })

    test("0", () => {
        let result: any = inst2.event("SELECT * FROM Movies WHERE Title=’Jurassic Park’ AND Director='Steven Spielberg';")
        expect(result).toMatchSnapshot()
    })

    test("1", () => {
        let result: any = inst.event("UPDATE Projects SET pname = %s WHERE pid = %s")
        expect(result).toMatchSnapshot()
    })

    test("2", () => {
        let result: any = inst3.event("DROP TABLE tmp;")
        expect(result).toMatchSnapshot()
    })

    test("3", () => {
        let result: any = inst4.event("UNLOCK TABLES;")
        expect(result).toMatchSnapshot()
    })

    test("4", () => {
        let result: any = inst5.event("")
        expect(result).toMatchSnapshot()
    })
})

describe("emitter.Emitter.size", () => {
    let inst2: any

    beforeEach(() => {
        inst2 = new emitter.Emitter()
    })

    test("0", () => {
        let result: any = inst2.size()
        expect(result).toMatchSnapshot()
    })
})
