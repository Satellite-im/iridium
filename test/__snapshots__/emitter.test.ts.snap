// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emitter.Emitter.emit 0 1`] = `Promise {}`;

exports[`emitter.Emitter.emit 1 1`] = `Array []`;

exports[`emitter.Emitter.emit 2 1`] = `Array []`;

exports[`emitter.Emitter.event 0 1`] = `Array []`;

exports[`emitter.Emitter.event 1 1`] = `Array []`;

exports[`emitter.Emitter.event 2 1`] = `Array []`;

exports[`emitter.Emitter.event 3 1`] = `Array []`;

exports[`emitter.Emitter.event 4 1`] = `Array []`;

exports[`emitter.Emitter.off 0 1`] = `Array []`;

exports[`emitter.Emitter.off 1 1`] = `Array []`;

exports[`emitter.Emitter.off 2 1`] = `Array []`;

exports[`emitter.Emitter.on 0 1`] = `1`;

exports[`emitter.Emitter.on 1 1`] = `1`;

exports[`emitter.Emitter.size 0 1`] = `0`;
